### PR TITLE
Add roslint verb

### PR DESCRIPTION
**Pull request by wkentaro**
https://github.com/ros/roslint/issues/48#issuecomment-162761102

### CODE REVIEW

**Comment by davetcoleman** [comment]
I'm happy to have this new alias, and we've been discussing roslint [here](https://github.com/wkentaro/catkin_tools/commit/66a15d2dcdcebceb9bd7fed4a3cd91e428dfe316), but I want to point out this alias throws errors if every package does not have the proper roslint dependencies added. What would more likely be the use case is using roslint per-package, such as:

    catkin build -v --this --no-deps --make-args roslint

Maybe that should be the proposed alias, with corresponding documentation?

**Comment by wkentaro** [comment]
hmm there is a same problem in `run_tests` doesn't it?
usually I use
```
catkin run_tests <package> --no-deps -iv
```
so for roslint
```
catkin roslint <package> --no-deps -iv
```

**Comment by NikolausDemmel** [comment]
Note there is also this discussion about roslint from a while back: https://github.com/catkin/catkin_tools/issues/201

> hmm there is a same problem in run_tests doesn't it?
usually I use

>     catkin run_tests <package> --no-deps -iv

> so for roslint

>     catkin roslint <package> --no-deps -iv

I'm fine with this alias analagous to `run_tests`, but maybe the docs should include the `catkin roslint <package> --no-deps` and `catkin roslint --this --no-deps` invocations as examples?

File: catkin_tools/config.py [hunk 1]
```diff
@@ -35,6 +35,7 @@
 install: config --install
 p: create pkg
 test: build --verbose --make-args test --
+roslint: build --verbose --make-args roslint --
```
Comment (davetcoleman) [inline catkin_tools/config.py]: Instead of just ``--verbose`` should it be ``-iv``?

**Comment by davetcoleman** [comment]
> but maybe the docs should include the catkin roslint <package> --no-deps and catkin roslint --this --no-deps invocations as examples?

+1

**Comment by wkentaro** [comment]
> but maybe the docs should include the catkin roslint --no-deps and catkin roslint --this --no-deps invocations as examples?

+1 good idea.

File: catkin_tools/config.py (hunk 1 above)
Comment (wkentaro) [inline catkin_tools/config.py]: I will change it.
the output seems better to see:
*before*
```
Starting ==> jsk_topic_tools                                                                                                                                           
                                                                                                                                                                       
[jsk_topic_tools] ==> '/home/wkentaro/ros/indigo/build/jsk_topic_tools/build_env.sh /usr/bin/make cmake_check_build_system' in '/home/wkentaro/ros/indigo/build/jsk_topic_tools'
[jsk_topic_tools] <== '/home/wkentaro/ros/indigo/build/jsk_topic_tools/build_env.sh /usr/bin/make cmake_check_build_system' finished with return code '0'
                                                                                                                                                                        
                                                                                                                                                                       
[jsk_topic_tools] ==> '/home/wkentaro/ros/indigo/build/jsk_topic_tools/build_env.sh /usr/bin/make roslint --jobserver-fds=3,5 -j' in '/home/wkentaro/ros/indigo/build/jsk_topic_tools'
Done processing src/log_utils.cpp
Total errors found: 0
Built target roslint_jsk_topic_tools
Built target roslint
[jsk_topic_tools] <== '/home/wkentaro/ros/indigo/build/jsk_topic_tools/build_env.sh /usr/bin/make roslint --jobserver-fds=3,5 -j' finished with return code '0'
                                                                                                                                                                        
Finished <== jsk_topic_tools [ 0.3 seconds ] 
```

*after*
```
Starting ==> jsk_topic_tools                                                                                                                                           
[jsk_topic_tools] ==> '/home/wkentaro/ros/indigo/build/jsk_topic_tools/build_env.sh /usr/bin/make cmake_check_build_system' in '/home/wkentaro/ros/indigo/build/jsk_topic_tools' 
[jsk_topic_tools] <== '/home/wkentaro/ros/indigo/build/jsk_topic_tools/build_env.sh /usr/bin/make cmake_check_build_system' finished with return code '0'              
[jsk_topic_tools] ==> '/home/wkentaro/ros/indigo/build/jsk_topic_tools/build_env.sh /usr/bin/make roslint --jobserver-fds=3,5 -j' in '/home/wkentaro/ros/indigo/build/jsk_topic_tools' 
[jsk_topic_tools] Done processing src/log_utils.cpp                                                                                                                    
[jsk_topic_tools] Total errors found: 0                                                                                                                                
[jsk_topic_tools] Built target roslint_jsk_topic_tools                                                                                                                 
[jsk_topic_tools] Built target roslint                                                                                                                                 
[jsk_topic_tools] <== '/home/wkentaro/ros/indigo/build/jsk_topic_tools/build_env.sh /usr/bin/make roslint --jobserver-fds=3,5 -j' finished with return code '0'        
Finished <== jsk_topic_tools [ 0.4 seconds ]       
```

Comment (wkentaro) [inline catkin_tools/config.py]: I hope `run_tests` also aliases with `-iv` option..

**Comment by wkentaro** [comment]
I updated the commits.

File: catkin_tools/config.py (hunk 1 above)
Comment (NikolausDemmel) [inline catkin_tools/config.py]: > I hope run_tests also aliases with -iv option..

I guess just `--verbose`. It's the line below where you added the roslint alias ;-).

**Comment by NikolausDemmel** [comment]
+1 for the updated version. Thanks.

**Comment by davetcoleman** [comment]
+1, now we need a review from @wjwwood or @jbohren

**Comment by wjwwood** [comment]
You guys are going to hate me, but I'd really rather not add an alias like this. I'd rather see a full verb which can intelligently run the linter over packages which actually use it rather than blindly trying to run that target on all packages. It could also do nice things like produce a summary of the linting results which is something the alias could never do. For me the purpose of the aliases is for individuals to make their work flow more convenient and the default aliases were just there to demonstrate how you might use them. I never intended or wanted people to contribute to the default aliases section.

I tried really hard to make sure you could externally add verbs so that we didn't have to merge things like this into the core `catkin_tools`. The reason for preferring externally added verbs is because then you don't have to wait for me to merge it and that way we don't have to re-release of `catkin_tools` before people can use it and it doesn't introduce any hard or soft dependencies into `catkin_tools`.

If the community really wants the alias then we can merge it, but I'd much prefer a verb. I'd also be interested in possibly exposing a way for external projects to contribute default aliases without changing the `catkin_tools` code (something like folder where anyone can drop alias fragments which `catkin_tools` can then pickup).

tl;dr

-0 to this alias being included. +1 to a verb that does this.

**Comment by davetcoleman** [comment]
:broken_heart:

While I agree creating a verb is a better long term solution, IMO this PR is a helpful step towards better integration or ROS tools for healthier code. Unless others want to put that effort in now, it seems this alias is good enough for our current needs.

However, I won't be too upset so long as we can just document this alias as something one could add themselves.

File: catkin_tools/config.py [hunk 2]
```diff
@@ -35,6 +35,7 @@
 install: config --install
 p: create pkg
 test: build --verbose --make-args test --
+roslint: build -iv --make-args roslint --
```
Comment (wjwwood) [inline catkin_tools/config.py]: Does the `roslint` target return a non-zero error code when there is a problem? If it does I would argue that there should be neither `-v` nor `-i`. I think those are preferences that the user should select. `-i` is particularly suspect and I'd say it shouldn't be the default even if `-v` is left.

File: docs/verbs/catkin_build.rst [hunk 3]
```diff
@@ -533,6 +533,32 @@ So if you want to check for failing tests, use the ``catkin_test_results`` comma
...
+
+.. code-block:: bash
+
+    $ catkin roslint
+
+Or the longer version:
+
+.. code-block:: bash
+
+    $ catkin build [...] --make-args roslint
```
Comment (wjwwood) [inline docs/verbs/catkin_build.rst]: Why not `--catkin-make-args`? There shouldn't be any non-catkin packages with the `roslint` target right?

File: catkin_tools/config.py (hunk 2 above)
Comment (davetcoleman) [inline catkin_tools/config.py]: Yes, for example I have 5 errors and I get:

```
[rviz_visual_tools] Total errors found: 5                                                                                                                                                                                                      
[rviz_visual_tools] make[3]: *** [roslint_rviz_visual_tools] Error 1                                                                                                                                                                           
[rviz_visual_tools] make[2]: *** [CMakeFiles/roslint_rviz_visual_tools.dir/all] Error 2                                                                                                                                                        
[rviz_visual_tools] make[1]: *** [CMakeFiles/roslint.dir/rule] Error 2                                                                                                                                                                         
[rviz_visual_tools] make: *** [roslint] Error 2                                                                                                                                                                                                
[rviz_visual_tools] <== '/home/dave/ros/current/ws_ros/build/rviz_visual_tools/build_env.sh /usr/bin/make roslint --jobserver-fds=3,5 -j' failed with return code '2'                                                                         
Failed   <== rviz_visual_tools [ 3.3 seconds ]                                                                                                                                                                                                 
[build] There were '1' errors:                                                                                                                                                                                                                 

Failed to build package 'rviz_visual_tools' because the following command:

# Command to reproduce:
cd /home/dave/ros/current/ws_ros/build/rviz_visual_tools && /home/dave/ros/current/ws_ros/build/rviz_visual_tools/build_env.sh /usr/bin/make roslint --jobserver-fds=3,5 -j; cd -

# Path to log:
cat /home/dave/ros/current/ws_ros/build/build_logs/rviz_visual_tools.log

Exited with return code: 2 
                                                                                                                                                                                                                                               
[build] Runtime: 3.4 seconds 
```

File: docs/verbs/catkin_build.rst (hunk 3 above)
Comment (davetcoleman) [inline docs/verbs/catkin_build.rst]: correct

**Comment by NikolausDemmel** [comment]
I guess the argument for a usable return code as in https://github.com/catkin/catkin_tools/issues/245 also applies here, necessitating implementing `catkin roslint` as a full verb.

I'm not sure what the downside of adding it as alias now, and replacing it with a verb later. I guess people might start relying on the specific behaviour. And I do agree that it is easy enough to add as an alias yourself if mentioned in the docs. Though I don't know if that is not worse, since once you add it as your own custom alias, it will not be removed when it is added as a full verb later and thus keep shadowing the verb implementation (or is this not how aliases work?).

**Comment by wjwwood** [comment]
An alias will override a verb. This is to allowed so that you can do something like `alias build=build -v`.

However, aliases are printed to the user every time, so the problem should be discoverable. But it's a good point. I'd argue that a minimal verb that just does what the alias is proposed to do would be a better starting place.

**Comment by wkentaro** [comment]
I think I updated the commit as your comments.
`roslint` should be verb or alias at this point?

**Comment by NikolausDemmel** [comment]
> I think I updated the commit as your comments.
Which roslint should be? verb or alias at this point?

I don't think there is really consensus, but I have to agree with @wjwwood that a minimal verb is probably the best way to go. IMHO in its simplest form it would invoke `make roslint` for the corresponding (catkin) packages (you could event check which ones actually depend on roslint). Any new verb could maybe best be built upon the upcoming https://github.com/catkin/catkin_tools/pull/249.

**Comment by wkentaro** [comment]
@NikolausDemmel thx for your comment.
I updated the commit and made `roslint` as verb not alias.
@wjwwood Could you please review this again?

**Comment by jbohren** [comment]
@wkentaro So I don't think `ros`-anything belongs in `catkin_tools`, including a `roslint` verb. What might make sense, however, is to add a `lint` verb, which uses `entry_points` so that others can provide linters for packages based on arbitrary tests.

**Comment by NikolausDemmel** [comment]
One could also package the roslint verb with the roslint package rather than with catkin_tools using entry points. For a general lint verb we'd need to gather some use cases to see how the interface might look and how they might interact or the results be combined.

**Comment by jbohren** [comment]
> One could also package the roslint verb with the roslint package rather than with catkin_tools using entry points. For a general lint verb we'd need to gather some use cases to see how the interface might look and how they might interact or the results be combined.

Yeah, I agree with @NikolausDemmel that regardless of how `catkin_tools` ends up organizing this functionality, this would best be added to the `roslint` package.

**Comment by jack-oquin** [comment]
Adding it to ``roslint`` entails checking which packages depend on that package. Not all of them do.

Unlike ``run_tests``, ``roslint`` is generally not a valid target for every package in a workspace.

**Comment by wjwwood** [comment]
I think the same is true for `run_tests` (not all packages will have this target, e.g. plain cmake packages). I think both the lint verb and the test verb could benefit from checking that the make file has that target, using this function for example: https://github.com/ros/catkin/blob/3b8a3b9607adff98ed32cfd3b6d6ea4aed647424/python/catkin/builder.py#L444-L454

For that matter the `--install` option should probably check first (or after a failure to give more information).

**Comment by wkentaro** [comment]
summarizing above,

- `roslint` verb should be included in `roslint` package
- `roslint` verb should check if `roslint` is the valid target of each package in a workspace

Am I right? If so I will close this PR and send PR to roslint package.

> @wkentaro So I don't think ros-anything belongs in catkin_tools, including a roslint verb. What might make sense, however, is to add a lint verb, which uses entry_points so that others can provide linters for packages based on arbitrary tests.

Actually, the `lint` verb is already added by `catkin_lint` tool.
https://github.com/fkie/catkin_lint/blob/master/setup.py#L39-L41

**Comment by jbohren** [comment]
> Am I right? If so I will close this PR and send PR to roslint package.

Yeah, that sounds reasonable.

> Actually, the `lint` verb is already added by catkin_lint tool.

We should start documenting these in a centralized place. #258

**Comment by wkentaro** [comment]
roslint is not a pure python package but a ros package.
In that case, how can I use `entry_points` in setup.py with `distutils`?
`setuptools` seems to offer that function..

- setup with distuils https://github.com/ros/roslint/blob/master/setup.py#L4-L9
- setup() with setuptools https://github.com/fkie/catkin_lint/blob/master/setup.py#L39-L41

**Comment by wjwwood** [comment]
@wkentaro that's a good point, I don't think I've ever done `entry_points` with catkin, not sure if that will work or not.

http://docs.ros.org/jade/api/catkin/html/user_guide/setup_dot_py.html seems to imply that `entry_points` are not support for the devel space, but they might work for the install space.

But this introduces a bootstrapping issue, which is that will have to build, install, and source `roslint` before using it's catkin_tools verb. I'm not sure if there is a more elegant way to do this.

**Comment by wkentaro** [comment]
As a progress about this issue, I have created below PRs:
- PR which makes catkin support entry_points of setup.py: https://github.com/ros/catkin/pull/798
- PR which depends on https://github.com/ros/catkin/pull/798, and registers roslint verb for catkin_tools as one of the entry_points

**Comment by wjwwood** [comment]
I'm going to close this pr in favor of either @wkentaro's work to support entry points in catkin, or even better (in my opinion), just have a roslint_catkin_tools_verb pure Python package that implements the verb outside of ROS.

I can reopen the pr in the future if we need to.
